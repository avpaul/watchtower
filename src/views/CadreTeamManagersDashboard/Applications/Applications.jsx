import React, { Component } from 'react';
import PropTypes from 'prop-types';
import weakKey from 'weak-key';
import PMloader from '../../../components/CustomLoader/PMLoader';
import TeamManagerCard from '../../../components/TeamManagerCard';
import { altDate as formatDate } from '../../../utils/formatDate';
import FilterDropdown from '../../../components/FilterDropdown';
import ApplicantCard from '../../../components/TeamManagerCard/ApplicantCard';

import './Applications.scss';

export default class Applications extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showApplication: false,
      application: {},
      filteredApplications: [],
      currentRole: 'All Roles'
    };
  }

  componentDidMount = async () => {
    const { fetchApplications } = this.props;
    try {
      await fetchApplications();
      const {
        applications: {
          data: { pending }
        }
      } = this.props;
      return this.setState({
        filteredApplications: pending
      });
    } catch (e) {
      return false;
    }
  };

  filterByRole = role => {
    const {
      applications: {
        data: { pending }
      }
    } = this.props;
    let filteredApplicants = pending.filter(
      applicant => applicant.role.name === role
    );
    if (filteredApplicants.length === 0) filteredApplicants = pending;
    return this.setState({
      currentRole: role,
      filteredApplications: filteredApplicants
    });
  };

  /**
   * Concatenates the first name and the last name
   * @param firstName
   * @param lastName
   */
  formatName = (firstName, lastName) => firstName.concat(' ', lastName);

  /**
   *
   * gets the applicants array and returns all roles
   * @param {object} applicants
   * @returns {array} roles
   */

  getRoles = applicants => {
    let roles = ['All Roles'];
    if (applicants.length > 0) {
      roles = applicants.reduce(
        (accumulator, application) => {
          const roleName = application.role.name;
          if (!accumulator.includes(roleName)) accumulator.push(roleName);
          return accumulator;
        },
        ['All Roles']
      );
    }
    return roles;
  };
  /**
   * Returns the page stats header
   * @param applications
   * @param filter
   * @param current
   */

  renderHeader = (applications, filter, current) => {
    const filterItems = this.getRoles(applications);
    return (
      <div className="stats-header mb-4 d-flex justify-content-between">
        <div>
          {applications.length > 0 ? applications.length : 0} Pending
          Applications
        </div>
        <FilterDropdown
          items={filterItems}
          current={current}
          width="400"
          chevronColor="#808FA3"
          dropdownBackgroundColor="#FFFFFF"
          getFilter={(type, value) => filter(value)}
        />
      </div>
    );
  };

  /**
   * Returns the card containing details about an applicant after being clicked
   * @param e
   * @param application
   */
  fetchSingleApplication = (e, application) => {
    const { showApplication } = this.state;
    this.setState({ showApplication: !showApplication, application });
  };

  mapTeamApplications = filteredApplications => {
    const teamApplications = filteredApplications.map(application => {
      const {
        role: { name: role },
        project,
        applicant: {
          picture,
          first_name: firstName,
          last_name: lastName,
          cohort,
          sims_start_date: startDate
        }
      } = application;
      return (
        <div className="mb-4" key={application.id}>
          <TeamManagerCard
            data-test="team_manager_card"
            role={role}
            event={e => this.fetchSingleApplication(e, application)}
            image={picture}
            key={weakKey(application)}
            name={this.formatName(firstName, lastName)}
            project={project.name}
            cohort={cohort}
            dateType="Start Date"
            date={formatDate(startDate)}
          />
        </div>
      );
    });
    return teamApplications.length ? (
      teamApplications
    ) : (
      <h1>There are currently no applications</h1>
    );
  };

  renderApplicantProfileCard = application => {
    const { showApplication } = this.state;
    return (
      <div
        className={`applicant_profile_card_wrapper ${
          showApplication ? 'show-application' : 'hide-application'
        }`}
      >
        <ApplicantCard application={application} />
      </div>
    );
  };

  render() {
    const {
      applications: { loading }
    } = this.props;

    const {
      showApplication,
      application,
      filteredApplications,
      currentRole
    } = this.state;

    return loading ? (
      <div className="cadre__page">
        <PMloader />
      </div>
    ) : (
      <div className="applications-wrapper">
        <div className="cadre__page">
          <div className="tm-project-applications">
            <div
              className={`all-applicants-cards d-flex flex-wrap justify-content-between ${
                showApplication ? 'add-width' : 'remove-width'
              }`}
            >
              {this.renderHeader(
                filteredApplications,
                this.filterByRole,
                currentRole
              )}
              {this.mapTeamApplications(filteredApplications)}
            </div>
          </div>
        </div>
        {showApplication && this.renderApplicantProfileCard(application)}
      </div>
    );
  }
}

Applications.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  applications: PropTypes.object.isRequired,
  fetchApplications: PropTypes.func.isRequired
};
