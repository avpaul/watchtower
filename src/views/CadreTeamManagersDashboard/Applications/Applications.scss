@import '../../../assets/sass/abstract/variables';
.tm-project-applications {
  width: 100%;
  .stats-header {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    color: #333e44;
    font-family: 'DIN Pro';
    font-size: 24px;
    letter-spacing: -1.1px;
    line-height: 31px;
    width: 100%;
    height: 100%;
  }
  .TM_ProfileCard {
    outline: none !important;
  }
}
.cadre__page {
  width: 100%;
}

.applicant_profile_card_wrapper {
  background: #ffffff;
  max-width: none !important;
  border-radius: 0 !important;
  margin-top: 0 !important;
  height: 100vh !important;
  width: 0;
  position: fixed;
  z-index: 1;
  top: 0;
  right: 0;
  overflow-x: hidden;
  transition: all 1s ease;
}

.show-application {
  width: 350px !important;
  padding: 150px 20px 10px;
  z-index: 1;
}

.hide-application {
  width: 0 !important;
}

.applications-wrapper {
  display: flex;
}

.all-applicants-card {
  transition: all 0.3s ease;
}

.add-width {
  width: 60%;
}

.remove-width {
  width: 100%;
}
