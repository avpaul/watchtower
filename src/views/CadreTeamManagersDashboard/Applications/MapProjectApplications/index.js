import MapProjectApplications from './MapProjectApplications';

export default MapProjectApplications;
