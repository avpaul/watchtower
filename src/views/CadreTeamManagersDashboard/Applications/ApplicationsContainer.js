import { connect } from 'react-redux';
import Applications from './Applications';
import { fetchApplications } from '../../../redux/actionCreators/cadreTeamManager/applicationsActions';
import { acceptApplication } from '../../../redux/actionCreators/cadreTeamManager/acceptApplicationActions';

export const mapStateToProps = ({ teamManagerProjectApplications }) => ({
  applications: teamManagerProjectApplications
});

export default connect(
  mapStateToProps,
  {
    fetchApplications,
    acceptApplication
  }
)(Applications);
