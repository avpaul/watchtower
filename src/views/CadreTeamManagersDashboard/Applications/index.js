import ApplicationsContainer from './ApplicationsContainer';

export default ApplicationsContainer;
