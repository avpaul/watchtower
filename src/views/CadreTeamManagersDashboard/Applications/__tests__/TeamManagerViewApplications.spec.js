import React from 'react';
import { mount } from 'enzyme';
import Applications from '../Applications';

let props;
let wrapper;

describe('Test team manager view applications', () => {
  it('should map display project instances properly', () => {
    props = {
      applications: {
        data: {
          pending: [
            {
              id: 16,
              role: {
                "id": 5,
                "name": "Data Engineer",
                "description": "Inventore eveniet sunt molestias reprehenderit.,Est culpa minima quod veritatis laudantium soluta suscipit.",
                "deleted_at": null,
                "duration": 3
              },
              project: {
                "id": 4,
                "name": "Omnis voluptatem quos."
              },
              applicant: {
                id: 9,
                fellow_id: '-HYBHVWK78HBJ',
                first_name: 'Test',
                last_name: 'Applicant',
                picture: 'https://lorempixel.com/100/100/people/?23475 ',
                cohort: 'NBO-11',
                sims_start_date: '2019-02-19 00:00:00'
              }
            }
          ],
          accepted: [],
          rejected: []
        }
      },
      fetchApplications: jest.fn()
    };
    wrapper = mount(<Applications {...props} />);
  });
  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
