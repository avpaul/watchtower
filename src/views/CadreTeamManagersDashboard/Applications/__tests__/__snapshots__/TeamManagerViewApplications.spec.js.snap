// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test team manager view applications should render correctly 1`] = `
<Applications
  applications={
    Object {
      "data": Object {
        "accepted": Array [],
        "pending": Array [
          Object {
            "applicant": Object {
              "cohort": "NBO-11",
              "fellow_id": "-HYBHVWK78HBJ",
              "first_name": "Test",
              "id": 9,
              "last_name": "Applicant",
              "picture": "https://lorempixel.com/100/100/people/?23475 ",
              "sims_start_date": "2019-02-19 00:00:00",
            },
            "id": 16,
            "project": Object {
              "id": 4,
              "name": "Omnis voluptatem quos.",
            },
            "role": Object {
              "deleted_at": null,
              "description": "Inventore eveniet sunt molestias reprehenderit.,Est culpa minima quod veritatis laudantium soluta suscipit.",
              "duration": 3,
              "id": 5,
              "name": "Data Engineer",
            },
          },
        ],
        "rejected": Array [],
      },
    }
  }
  fetchApplications={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="applications-wrapper"
  >
    <div
      className="cadre__page application__page"
    >
      <div
        className="tm-project-applications"
      >
        <div
          className="all-applicants-cards d-flex flex-wrap remove-width"
        >
          <div
            className="stats-header mb-4 d-flex justify-content-between"
          >
            <div>
              0
               Pending Applications
            </div>
            <Filter
              characterLength={9}
              chevronColor="#808FA3"
              current="All Roles"
              dropdownBackgroundColor="#FFFFFF"
              fontSize=""
              getFilter={[Function]}
              items={
                Array [
                  "All Roles",
                ]
              }
              search={false}
              width="400"
            >
              <div
                className="filter-dropdown"
              >
                <button
                  className="filter-dropdown__button"
                  onClick={[Function]}
                  style={
                    Object {
                      "fontSize": "",
                      "width": "100%",
                    }
                  }
                  type="button"
                >
                  <span>
                    All Roles
                  </span>
                  <span
                    style={
                      Object {
                        "background": "#FFFFFF",
                      }
                    }
                  >
                    <i
                      className="fas fa-chevron-down"
                      style={
                        Object {
                          "color": "#808FA3",
                        }
                      }
                    />
                  </span>
                </button>
              </div>
            </Filter>
          </div>
          <h1>
            There are currently no applications
          </h1>
        </div>
      </div>
    </div>
  </div>
</Applications>
`;
